### PR TITLE
Improve mingw includes utilizing ```-iquote``` option to simply include headers from any custom directory

### PR DESCRIPTION
**Pull request by araout42**
## Overview

this PR is linked to the issue #17588 
These changes improve the Mingw compiler to allow to include headers from custom dirs, and differenciate it from the Mingw native includes 
it utilize the option [-iquote](https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html)
 
So for an example a mingw.rb consumer could include from this path 
```File.join(Msf::Config.data_directory, "headers", "windows")```
and at the same time do  ```#include <stdio.h>``` from the native mingw include dir without any conflict

## Verification

create an example file ```test.rb``` at the path ```modules/evasion/windows/```
with the given content 

```
require 'metasploit/framework/compiler/windows'

class MetasploitModule < Msf::Evasion

  def initialize(info={})
    super(merge_info(info,
      'Name'        => 'test',
      'Targets'     => [ ['Microsoft Windows', {}] ],
      'Platform'    => 'win',
      'Arch'        => ARCH_X86,
    ))
  end
  

  def c_template
    @c_template ||= %Q|
#include <windows.h>
#include <stdio.h>
#include "rc4.h"


int main(int argc, char **argv)
{
  char str[] = "Hello World!";
  printf("%s\\n", str);
  return 0;
}|
  end

  def run
    dir = [File.join(Msf::Config.data_directory, "headers", "windows"), "/tmp/inc", "/home/my_custom/include_path"]
    m = Metasploit::Framework::Compiler::Mingw::X86.new({ show_compile_cmd: true, f_name: "OK", compile_options: " -lpsapi -lwininet ", include_dirs: dir })
    bin = m.compile_c(c_template)
    print_status("#{bin}")
  end
end

```

- [ ] Start `msfconsole`
- [ ] `use evasion/windows/test`
- [ ] `set payload windows/meterpreter/reverse_tcp`
- [ ] `set lhost 0.0.0.0`
- [ ] `exploit`

Verify that the file has been compiled correctly and no error output is given
the non-existing include dirs ```"/tmp/inc", "/home/my_custom/include_path"``` given in the options ``` include_dirs: dir```
should also not prevent the compilation

## more
of course when pushing a new module to the  repo use of the  ```include_dirs``` options it a bad practice to include path that are not already in the metasploit-framework repo 

But the ability to  include from any local directory can be useful anyway

### CODE REVIEW

**Comment by dwelch-r7** [comment]
For testers:
Pay close attention and validate encrypted payload generation and bof loader support

File: lib/metasploit/framework/compiler/mingw.rb [hunk 1]
```diff
@@ -34,6 +34,9 @@ def build_cmd(src)
 
           cmd << "#{self.mingw_bin} "
           cmd << "#{src_file} -I #{INCLUDE_DIR} "
+          if self.include_dirs
+            cmd << " #{self.include_dirs} "
+          end
```
Comment (adfoster-r7) [inline lib/metasploit/framework/compiler/mingw.rb:39]:
```suggestion
         cmd << "#{self.include_dirs.map { |include_dir| "-iquote #{include_dir}" }.join(' ')} " if self.include_dirs.any?
```

**Review by adfoster-r7 — COMMENTED** [review]
Small suggestion; Haven't tested - but it looks like it should be semantically equivalent 🤞 The intended goal was to allow folk to interact with `include_dirs` programatically in the future - without it being mixed up with the semantics of string option names such as `iquote` etc

Let me know if the suggestion works for you and we should be able to get this landed pretty shortly after

**Comment by araout42** [comment]
Hey, I like the proposition you made, your way looks better to me

I will review these changes and test them tomorrow

I will comment about my testings

And tell you if everything is good for me 

Thanks !

File: lib/metasploit/framework/compiler/mingw.rb (hunk 1 above)
Comment (araout42) [reply lib/metasploit/framework/compiler/mingw.rb:39]: I'm not sure about the #{include_dir} I think the 's' of `include_dirs` misses

**Comment by araout42** [comment]
I tested, and reviewed the review 😎
Everything works fine i add the followint sreenshots of my testings 

##**The script ** 
image


Gives this command line as a result : 
image


Wich is absolutely correct !
Everything is Good to me and can be merged, i have nothing to add 

Just one thing i would like to re-mention 
No one must push to the repo with an external directory given as iquote, But it's absolutely possible to push to the repo with  iquote reference to any directory of the repo

Thank you

File: lib/metasploit/framework/compiler/mingw.rb [hunk 2]
```diff
@@ -34,6 +34,9 @@ def build_cmd(src)
 
           cmd << "#{self.mingw_bin} "
           cmd << "#{src_file} -I #{INCLUDE_DIR} "
+          if self.include_dirs
+            cmd << " #{self.include_dirs} "
+         cmd << "#{self.include_dirs.map { |include_dir| "-iquote #{include_dir}" }.join(' ')} " if self.include_dirs.any?
```
Comment (adfoster-r7) [inline lib/metasploit/framework/compiler/mingw.rb:39]: Looks like Github has made a small mistake here; I think this should be:

```suggestion
          cmd << "#{self.include_dirs.map { |include_dir| "-iquote #{include_dir}" }.join(' ')} " if self.include_dirs.any?
```

Comment (araout42) [reply lib/metasploit/framework/compiler/mingw.rb:39]: Yes ! thanks

**Comment by adfoster-r7** [comment]
## Release Notes

Adds support for module writers to supply a custom `include_dirs` array when using the MinGW library to compile payloads

**Comment by adfoster-r7** [comment]
Thanks! :+1:

**Pull request merged**